python/main_loop.py

```diff
@@ -148,12 +154,19 @@ def createNewLocalPath(sailbot, maxAllowableRuntimeSeconds, desiredHeadingPublis
                 sailbot.newGlobalPathReceived = False
 
             # If globalWaypoint reached, increment the index
-            if isGlobalWaypointReached or goalInvalid:
-                if isGlobalWaypointReached:
-                    rospy.loginfo("Global waypoint reached")
-                if goalInvalid:
-                    rospy.logwarn("Goal state invalid, skipping to the next global waypoint")
+            if isGlobalWaypointReached:
+                rospy.loginfo("Global waypoint reached")
                 sailbot.globalPathIndex += 1
+                goalWasInvalid = False
+
+            # If goal invalid, move global waypoint towards sailbot until it is valid
+            if goalInvalid and not goalWasInvalid:
```

Comment:
I like your solution, it seems to be working better. BUT we still have the issue that the local path visualizer is a bit confusing. 

I realized that the function moveGlobalWaypointUntilValid() only depends on state. this whole update thing depends on both state and also the global variable goalWasInvalid. Could you look into putting this functionality all into sailbot class?

Then both main_loop and local_path_visualizer would use the sailbot class the same way and have it work. Eg. update sailbot.getCurrentState to automatically do this obstacle check and move of global waypoint. May also need to have some sort of class variable to be the equivalent of "goalWasInvalid" in sailbot. Let me know if that's unclear.

Pseudocode:


state = sailbot.getCurrentState()  
# This currently just gets current state.
# Now can get state, checks if goal latlon is invalid. If valid, then just use that. If not valid, then check if was previously invalid. If so, use that previous invalid latlon. If not previously invalid, then move global waypoint until valid and store this.

Can be used in local path visualizer as well to make both consistent.